include/pybind11/common.h

```diff
@@ -328,18 +328,47 @@ inline static constexpr int log2(size_t n, int k = 0) { return (n <= 1) ? k : lo
 
 inline std::string error_string();
 
-/// Core part of the 'instance' type which POD (needed to be able to use 'offsetof')
-template <typename type> struct instance_essentials {
+/** The space to allocate for simple layout instance holders (see below) in multiple of the size of
+ * a pointer (e.g.  2 means 16 bytes on 64-bit architectures).  The default is the minimum required
+ * to holder either a std::unique_ptr or std::shared_ptr (which is almost always
+ * sizeof(std::shared_ptr<T>)).
+ */
+constexpr size_t instance_simple_holder_in_ptrs() {
+    static_assert(sizeof(std::shared_ptr<int>) >= sizeof(std::unique_ptr<int>),
+            "pybind assumes std::shared_ptrs are at least as big as std::unique_ptrs");
+    return (sizeof(std::shared_ptr<int>) / sizeof(void *)) + (sizeof(std::shared_ptr<int>) % sizeof(void *) != 0);
+}
+
+/// The 'instance' type which needs to be standard layout (needed to be able to use 'offsetof')
+struct instance {
     PyObject_HEAD
-    type *value;
+    // Storage for pointers and holder; see simple_layout, below, for a description
+    union {
+        void *simple_value_holder[1 + instance_simple_holder_in_ptrs()];
+        void **values_and_holders;
+    };
     PyObject *weakrefs;
+    /// If true, the pointer is owned which means we're free to manage it with a holder.
     bool owned : 1;
-    bool holder_constructed : 1;
-};
-
-/// PyObject wrapper around generic types, includes a special holder type that is responsible for lifetime management
-template <typename type, typename holder_type = std::unique_ptr<type>> struct instance : instance_essentials<type> {
-    holder_type holder;
+    /** An instance has two possible value/holder layouts.
+     *
+     * Simple layout (when this flag is true), means the `simple_value_holder` is set with a pointer
+     * and the holder object governing that pointer, i.e. [val1*][holder].  This layout is applied
+     * whenever there is no python-side multiple inheritance of bound C++ types *and* the type's
+     * holder will fit in the default space (which is large enough to hold either a std::unique_ptr
+     * or std::shared_ptr).
+     *
+     * Non-simple layout applies when using custom holders that require more space than `shared_ptr`
+     * (which is typically the size of two pointers), or when multiple inheritance is used on the
+     * python side.  Non-simple layout allocates the required amount of memory to have multiple
+     * bound C++ classes as parents.  Under this layout, `values_and_holders` is set to a pointer to
+     * allocated space of the required space to hold a holder-constructed bitset followed by a
+     * sequence of value pointers and holders, i.e. [bbb...][val1*][holder1][val2*][holder2]...
+     * where each [block] is rounded up to a multiple of `sizeof(void *)`.
```

Comment:
That encapsulation is exactly what the `value_and_holder` struct is for.  As such, only three places right now deal with the layout: `class_support.h` new/dealloc, the allocation code in `cast.h`'s many-argument `cast()` (which doesn't use new to avoid the unwanted instance pointer allocation, but also can be simpler because it knows there is only one base class); and the `get_value_and_holder()` logic itself.  No other code should ever need to touch it.